File: blind_75_leetcode/python/s0020_valid_parentheses.py
import unittest


class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) & 1:
            return False
        stack = []
        parentheses = {"(": ")", "[": "]", "{": "}"}
        for c in s:
            if c in parentheses:
                stack.append(c)
            else:
                if not stack or parentheses[stack.pop()] != c:
                    return False

        return not stack


class TestS20(unittest.TestCase):
    def test_works(self):
        s = Solution()
        self.assertEqual(s.isValid("()"), True)
        self.assertEqual(s.isValid("()[]{}"), True)
        self.assertEqual(s.isValid("(]"), False)
        self.assertEqual(s.isValid("}("), False)

File: blind_75_leetcode/src/solution/s0020_valid_parentheses.rs
#![allow(dead_code)]
pub struct Solution {}

impl Solution {
    pub fn is_valid(s: String) -> bool {
        if s.len() & 1 == 1 {
            return false;
        }
        let mut stack = Vec::with_capacity(s.len());
        for c in s.chars() {
            match c {
                ')' | ']' | '}' => {
                    if let Some(c1) = stack.pop() {
                        if !((c == ')' && c1 == '(')
                            || (c == ']' && c1 == '[')
                            || (c == '}' && c1 == '{'))
                        {
                            return false;
                        }
                    } else {
                        return false;
                    }
                }
                _ => stack.push(c),
            }
        }
        stack.is_empty()
    }
}

#[cfg(test)]
mod tests_20 {
    use super::*;

    #[test]
    fn it_works() {
        assert_eq!(Solution::is_valid("()".to_string()), true);
        assert_eq!(Solution::is_valid("(]".to_string()), false);
        assert_eq!(Solution::is_valid("()[]{}".to_string()), true);
    }
}
